### PR TITLE
Propagate generator's cancellation token to Parallel.For

### PR DESCRIPTION
**Pull request by captainsafia**

### CODE REVIEW

File: src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs [hunk 1]
```diff
@@ -257,13 +257,15 @@ private static string GetIdentifierFromPath(string filePath)
...
+        private static ParallelOptions GetParallelOptions(GeneratorExecutionContext generatorExecutionContext)
         {
             if (Debugger.IsAttached)
             {
                 return new ParallelOptions { MaxDegreeOfParallelism = 1 };
             }
-            return DefaultParallelOptions;
+            var options =  DefaultParallelOptions;
+            options.CancellationToken = generatorExecutionContext.CancellationToken;
+            return options;
```
Comment (pranavkm) [inline src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs:268]: We should use a non-static instance if we're mutating this:

```suggestion
            var options = new ParallelOptions { CancellationToken = generatorExecutionContext.CancellationToken };
```

File: src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs [hunk 2]
```diff
@@ -257,13 +257,14 @@ private static string GetIdentifierFromPath(string filePath)
...
 
-        private static ParallelOptions GetParallelOptions()
+        private static ParallelOptions GetParallelOptions(GeneratorExecutionContext generatorExecutionContext)
         {
             if (Debugger.IsAttached)
             {
                 return new ParallelOptions { MaxDegreeOfParallelism = 1 };
             }
-            return DefaultParallelOptions;
+            var options = new ParallelOptions { CancellationToken = generatorExecutionContext.CancellationToken };
```
Comment (pranavkm) [inline src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs:266]: Could you also remove `DefaultParallelOptions`?

**Review by pranavkm — APPROVED** [review]

File: src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs [hunk 3]
```diff
@@ -257,13 +255,14 @@ private static string GetIdentifierFromPath(string filePath)
             return builder.ToString();
         }
 
-        private static ParallelOptions GetParallelOptions()
+        private static ParallelOptions GetParallelOptions(GeneratorExecutionContext generatorExecutionContext)
         {
             if (Debugger.IsAttached)
             {
                 return new ParallelOptions { MaxDegreeOfParallelism = 1 };
             }
```
Comment (jaredpar) [inline src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs:263]: Seeing this reminded me of another change we should make. The compiler parallelism can be explicitly disabled by users. This is a flag that generators should respect. Essentially if the user demands the compiler run in single threaded mode then generators loaded in the compiler should respect that. 

In the case that [CompilationOptions.ConcurrentBuild](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Compilation/CompilationOptions.cs,ae09d3a82430aa43,references) is disabled I think we should have a `MaxDegreeOfParallelism` of one here or better avoid the `Parallel.For` entirely. 

Can take this up in a separate change though, just thought about it while reading this.

Comment (captainsafia) [reply src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs:263]: Good tip -- I can cook up a follow-up PR to resolve this.

**Comment by captainsafia** [comment]
I'm not upset, bot, just disappointed.

**Pull request merged**